packages/teleterm/src/ui/uri.ts

```diff
@@ -30,6 +30,54 @@ export const paths = {
   doc: '/docs/:docId',
 };
 
+type RootClusterId = string;
+type LeafClusterId = string;
+type ServerId = string;
+type KubeId = string;
+type DbId = string;
+export type RootClusterUri = `/clusters/${RootClusterId}`;
+export type RootClusterServerUri =
+  `/clusters/${RootClusterId}/servers/${ServerId}`;
```

Comment:
I mean, technically the current implementation would at least prohibit us from passing a root cluster URI where a leaf cluster URI is expected. But I don't think this is that useful, we don't have many places where leaf cluster URIs are allowed but root cluster URIs are prohibited. And it doesn't outweigh the false positive of passing leaf cluster URI as a root cluster URI.

Reply:
>It would be great if we could restrict RootClusterId to a stricter type then string.

AFAIK it is not possible, hopefully in the future we will be able to create [regex-validated string types](https://github.com/microsoft/TypeScript/issues/41160).

>If it's not possible, then perhaps we shouldn't define types which cannot be enforced and instead focus just on those that we actually can.

Hmm, I would define them anyway. They would act as [type aliases](https://www.typescriptlang.org/docs/handbook/advanced-types.html#type-aliases). I think it is still better for the reader to see `clusterUri: RootClusterUri` than `clusterUri: string`. You could also add a short comment to the type definition with an example.

Reply:
I don't think our types need to differentiate between root/leaf (i mean... if it was an easy fix sure), but as you said, there is nowhere that expects _leaf_ only, just places that require root only. 

I agree that defining them anyway would help future readers. not only would we have params named stuff like `rootClusterUri` but also typed for double safety.

Reply:
> I think it is still better for the reader to see `clusterUri: RootClusterUri` than `clusterUri: string`.

Ah, those would still be typed, we wouldn't use `string`. It's just that because we cannot differentiate between root and leaf URIs, we'd have a single `ClusterUri` type that accepts both. So instead of `clusterUri: RootClusterUri` you'd have `rootClusterUri: ClusterUri`. And as Michael said, I think in those cases the name of the variable should be `rootClusterUri` no matter what the type is. ;)

---

However, we could make the same argument about cluster URIs vs resource URIs. This code works:

```typescript
const serverUri: RootClusterServerUri = '/clusters/foo/servers/bar'
const uri: RootClusterUri = serverUri
const leafServerUri: LeafClusterServerUri = '/clusters/foo/leaves/baz/servers/bar'
const leafUri: LeafClusterUri = leafServerUri
```

But I wouldn't want to remove the distinction between clusters and resources in types simply because we can't prevent resource URIs being passed as cluster URIs.

So yeah, maybe let's keep it as it is for now and see how it works out. Refactoring this will be super easy anyway.